### PR TITLE
feat: PlanarFreehandROI stats

### PR DESCRIPTION
**Pull request by m00n620**
Display PlanarFreehandROI stats like other ROI tools (Elliptical, Rectangle).

### CODE REVIEW

File: packages/tools/src/tools/annotation/PlanarFreehandROITool.ts [hunk 1]
```diff
@@ -205,6 +213,12 @@ class PlanarFreehandROITool extends AnnotationTool {
     registerOpenContourEditLoop(this);
     registerOpenContourEndEditLoop(this);
     registerRenderMethods(this);
+
+    this._throttledCalculateCachedStats = throttle(
```
Comment (sedghi) [inline packages/tools/src/tools/annotation/PlanarFreehandROITool.ts:217]: can you add a configuration to have the cached stats on or off and I guess default to off?

Comment (m00n620) [reply packages/tools/src/tools/annotation/PlanarFreehandROITool.ts]: @sedghi , please take a look again. `calculateStats` configuration varaible has been added.

**Comment by m00n620** [comment]
@lscoder, can you please review this as soon as possible?

**Comment by lscoder** [comment]
@m00n620 Users should be able to drag and drop labels but it is broken

File: packages/tools/src/tools/annotation/PlanarFreehandROITool.ts [hunk 2]
```diff
@@ -568,11 +580,81 @@ class PlanarFreehandROITool extends AnnotationTool {
...
       // render contour method to render all of them
-      annotations.forEach((annotation) =>
-        this.renderContour(enabledElement, svgDrawingHelper, annotation)
-      );
+      annotations.forEach((annotation) => {
+        this.renderContour(enabledElement, svgDrawingHelper, annotation);
+
+        if (!this.configuration.calculateStats) return;
+
+        const { data } = annotation;
```
Comment (lscoder) [inline packages/tools/src/tools/annotation/PlanarFreehandROITool.ts:592]: Could you also please move all the new code below into a new function (eg: `renderStats` or something like that) leaving only two functions calls in the `forEach` loop. That helps having smaller and more specific functions that are easier to maintain.

**Comment by m00n620** [comment]
> lscoder

@lscoder , it requires heavy calculation for stats, so I didn't implement moving stats, should we essentially need this feature?

File: packages/tools/src/tools/annotation/PlanarFreehandROITool.ts (hunk 2 above)
Comment (m00n620) [reply packages/tools/src/tools/annotation/PlanarFreehandROITool.ts:611]: This is resolved, @lscoder

**Comment by lscoder** [comment]
> > lscoder
> 
> @lscoder , it requires heavy calculation for stats, so I didn't implement moving stats, should we essentially need this feature?

All labels work in the same way and we follow the same standard to all tools. Why does it need heavy calculation when moving the label? The points should be the same and it could use the stats from the cache.

**Comment by m00n620** [comment]
@lscoder , @sedghi , moving textbox feature has been added. please take a look again.

**Comment by sedghi** [comment]
There are some improvements to this PR, but not there yet
- textboxes are removed when drawing happens that means all textboxes are recalculated?
- Drawing on the second viewport make the textbox on the first one to disappear

[cb22d03c-1614-4dbc-be79-57d93e791cdf.webm](https://user-images.githubusercontent.com/7490180/217615077-d0ca9f37-f0b3-40da-8960-c8dd4632214a.webm)

**Comment by wayfarer3130** [comment]
Please ensure that you have tested store/remember with the @cornerstonejs/adapters DICOM SR reader and writer.  Those should be able to record and restore the Planar Freehand object with the updated measurements data and position.

**Comment by m00n620** [comment]
> There are some improvements to this PR, but not there yet
> 
> * textboxes are removed when drawing happens that means all textboxes are recalculated?
> * Drawing on the second viewport make the textbox on the first one to disappear
> 
>  cb22d03c-1614-4dbc-be79-57d93e791cdf.webm

@sedghi , this comment has been resolved, please take a look again.

**Comment by m00n620** [comment]
> Please ensure that you have tested store/remember with the @cornerstonejs/adapters DICOM SR reader and writer. Those should be able to record and restore the Planar Freehand object with the updated measurements data and position.

@wayfarer3130 , can you please be more specific? it would be great if you can comment more for how to test storing/remembering with the @cornerstonejs/adapters DICOM SR reader and writer on this tool.

**Comment by sedghi** [comment]
@m00n620 @wayfarer3130 We can focus on performance for stats for this PR and deal with SR later.
@m00n620 I see a lot of performance issue on this tool, have you tried it locally? 
This demo is basically not usable https://deploy-preview-326--cornerstone-3d-docs.netlify.app/live-examples/planarfreehandroitool

**Comment by m00n620** [comment]
> 

@sedghi , can you please be more specific about performance issues? I tried locally and I think it's working well.

**Comment by sedghi** [comment]
Can you send me a video capturing the link I sent you when you draw a contour and enable text box and enable again?

**Comment by m00n620** [comment]
@sedghi , I found out that there are some performance issues when I draw large-area contours. I will do more adjustments and let you know when it's ready.

**Comment by wayfarer3130** [comment]
> > Please ensure that you have tested store/remember with the @cornerstonejs/adapters DICOM SR reader and writer. Those should be able to record and restore the Planar Freehand object with the updated measurements data and position.
> 
> @wayfarer3130 , can you please be more specific? it would be great if you can comment more for how to test storing/remembering with the @cornerstonejs/adapters DICOM SR reader and writer on this tool.

You need to embed this within a framework such as OHIF that has a store/remember facility.  I'd like to add this to the CS3D with a set of examples, but that isn't available yet.  I think proceeding without the adapters is probably worthwhile, as the save/load can then be added separately.

**Comment by m00n620** [comment]
> @m00n620 It seems stats are being recalculated when user draws new ROIs even when it is not needed because it gets slower considerably. After comparing to the [video](https://github.com/cornerstonejs/cornerstone3D-beta/pull/326#issuecomment-1423035571) shared by Alireza the textboxes were being removed but not recalculated in the middle of the drawing process and now they are not being removed but recalculated on every mouse move event.
> 
>  2023-04-13.11-01-11.mp4 
> 1st ROI - OK 2nd ROI - slower 3rd ROI - hard to draw

@lscoder , this should be fixed now. please check again.

**Review by wayfarer3130 — APPROVED** [review]
@sedghi - I've tested the various cases that were complained about, and also looked at the various comments, and I think they are all addressed.  If you are ok with that, then I can merge this.  I think it is sufficiently ready to be worth including now - not perfect, but better than before.

**Comment by m00n620** [comment]
@lscoder , can you please take a look again?

**Review by shivasuri — APPROVED** [review]

**Comment by lscoder** [comment]
LGTM

**Review by lscoder — APPROVED** [review]

**Comment by rghgit1** [comment]
> OHIF

Have been looking for PlanarFreehandROI stats in OHIF Viewer V3. Has this been integrated with OHIF Viewer? If not, can you please guide me with the steps to do the same. Thanks

**Comment by m00n620** [comment]
> > OHIF
> 
> Have been looking for PlanarFreehandROI stats in OHIF Viewer V3. Has this been integrated with OHIF Viewer? If not, can you please guide me with the steps to do the same. Thanks
> 
> Adding some additional details on the steps tried so far:
> 
> 1. Updated the dependencies to latest versions of cornerstonejs
>    "@cornerstonejs/core": "^0.46.1",
>    "@cornerstonejs/tools": "^0.67.1",
> 2. Ran 'yarn install'
> 3. The code changes done as part of this feature is now available under extensions\cornerstone\node_modules@cornerstonejs\tools\src\tools\annotation\PlanarFreehandROITool.ts
> 
> However, this is not getting called when the PlanarFreehandROI tool is used and no stats are displayed in OHIF Viewport. Any pointers on how to invoke this when PlanarFreehandROI tool is used in OHIF would be great. Thanks

@rghgit1 , you should set `calculateStats` option in the configuration.

**Comment by rghgit1** [comment]
> > > OHIF
> > 
> > 
> > Have been looking for PlanarFreehandROI stats in OHIF Viewer V3. Has this been integrated with OHIF Viewer? If not, can you please guide me with the steps to do the same. Thanks
> > Adding some additional details on the steps tried so far:
> > 
> > 1. Updated the dependencies to latest versions of cornerstonejs
> >    "@cornerstonejs/core": "^0.46.1",
> >    "@cornerstonejs/tools": "^0.67.1",
> > 2. Ran 'yarn install'
> > 3. The code changes done as part of this feature is now available under extensions\cornerstone\node_modules@cornerstonejs\tools\src\tools\annotation\PlanarFreehandROITool.ts
> > 
> > However, this is not getting called when the PlanarFreehandROI tool is used and no stats are displayed in OHIF Viewport. Any pointers on how to invoke this when PlanarFreehandROI tool is used in OHIF would be great. Thanks
> 
> @rghgit1 , you should set `calculateStats` option in the configuration.

Thanks a ton. I had set `calculateStats` option to true, but was not at the right place. After setting `calculateStats` to 'true' in toolsConfig for PlanarFreehandROI, stats are getting displayed.

**Comment by rghgit1** [comment]
For certain scans, only Area is getting calculated. Max is shown as Infinity. Mean and Std Dev are not displayed. Is this a knwon issue?

image

**Pull request merged**